Support for OS("")

alexsaveliev opened the pull request:
This is how I see #3  :smile: 

- we do this so on Windows we can have a VFS which allows access to /C:/foo, /D:/foo, etc
- added test case meant to be run on Windows
- bonus: ability to run Travis CI builds on forked repository
- bonus: Appveyor integration

alexsaveliev committed:
Support for OS("")

- we do this so on Windows we can have a VFS which allows access to /C:/foo, /D:/foo, etc
- added test case meant to be run on Windows
- bonus: ability to run Travis CI builds on forked repository
- bonus: Appveyor integration

keegancsmith commented on os_windows.go:
```diff
@@ -0,0 +1,28 @@
+	// If it did, the result of filepath.Join would be outside the
+	// tree rooted at root.  We probably won't ever see a path
+	// with .. in it, but be safe anyway.
+	path = strings.TrimPrefix(filepath.Clean(filepath.Join(string(root), path)), string(os.PathSeparator))
```
this code allows you to escape `root`. Say root is `C:/foo` and path is `../bar`. When this line run path will be `c:/bar`.

alexsaveliev replied:
fixed

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,120 @@
+	// keys are mount points,
+	// values indicate if OS.Stat(path) should succeed for a given path
+	tests := map[string]map[string]bool{
+		empty: {
```
can you just write the strings here instead? I think the test case is easier to understand that way.

alexsaveliev replied:
done

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,120 @@
+			}
+
+		}
+	}
```
I think you could make this a unit test by removing the tmp file + os operations and instead just testing the resolve function on windows. That would probably make this test much easier to understand since you can use simple hardcoded paths.

alexsaveliev replied:
done

keegancsmith reviewed (changes requested):
This change is great. It still feels a little strange to special case "", but I'm keen to go ahead with it. Something with less special edge cases is to instead create a new VFS called `ctxvfs.RootOS`. However, this works!

I have requested changes because the test case is hard to read as it is, so I haven't reviewed that closely to understand if we are getting the desired behaviour.

alexsaveliev committed:
Taking into account code review comments

- disabled possibility to access files outside of mount point
- rewrote tests to use `path => expected` map
- removed local FS access in tests

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		"C:": {
+			"C:/folder/file":  "C:\\folder\\file",
+			"/C:/folder/file": "C:\\folder\\file",
```
why are you allowed to specify `C:/` and `/C:/`. Surely you should only be able to specify paths relative to `C:`? I expect the above to cases to resolve to the empty string.

Does the old `vfs.OS("C:")` have this behaviour? This crops up a few times in the test case, I'll just simply comment on them, but refer to this inline comment.

alexsaveliev replied:
> why are you allowed to specify C:/ and /C:/. Surely you should only be able to specify paths relative to C:? I expect the above to cases to resolve to the empty string.

I believe we should allow anything that located under the given mount point when combined with `root`. With the old implementation on Unix we can resolve "/foo/bar" to /"foo/bar" with mount point defined as a "/foo" so why do it different way on Windows?

>  Does the old vfs.OS("C:") have this behaviour? 

It doesn't. However on Unix old implementation gives the following results for `osFS.resolve` using `/foo` as a mount point:

"/foo/bar" =>  "/foo/bar", "bar =>  "/foo/bar"

I think that we should support both absolute and relative paths while they still located under a given mount point (or located system-wide in the cases of `/` or `""`)

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+			"/file":           "C:\\file",
+			"D:/folder/file":  "",
+		},
+		"/C:": {
```
Does `vfs.OS` work with a root path like `/C:` on windows

alexsaveliev replied:
Nope, old `vfs.OS` doesn't work with `/C:`. With the old implementation:

```
got `\C:\foo` while resolving `foo` on `/C:`
got `\C:\foo` while resolving `/foo` on `/C:`
```
which cannot be translated to real OS path when trying to access file content

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		"/C:": {
+			"C:/folder/file":  "C:\\folder\\file",
+			"/C:/folder/file": "C:\\folder\\file",
```
Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		"C:/folder": {
+			"C:/folder/file":           "C:\\folder\\file",
+			"/C:/folder/file":          "C:\\folder\\file",
```
Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+			"/C:\\anoherfolder/file":   "",
+			"foo/bar/../../file":       "C:\\folder\\file",
+			"foo/bar\\\\\\..\\../file": "C:\\folder\\file",
+			"c:/Folder/fiLe":           "c:\\Folder\\fiLe",
```
Above line same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		"/C:/folder": {
+			"C:/folder/file":         "C:\\folder\\file",
+			"/C:/folder/file":        "C:\\folder\\file",
```
Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

keegancsmith commented on os_windows_test.go:
```diff
@@ -0,0 +1,82 @@
+		},
+		"C:/folder/": {
+			"C:/folder/file":         "C:\\folder\\file",
+			"/C:/folder/file":        "C:\\folder\\file",
```
Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508
